dagql/idproto/id.proto

```diff
@@ -0,0 +1,98 @@
+syntax = "proto3";
+package dagger;
+
+option go_package = "./idproto";
+
+// ID represents a GraphQL value of a certain type, constructed by evaluating
+// its contained pipeline. In other words, it represents a
+// constructor-addressed value, which may be an object, an array, or a scalar
+// value.
+//
+// It may be binary=>base64-encoded to be used as a GraphQL ID value for
+// objects. Alternatively it may be stored in a database and referred to via an
+// RFC-6920 ni://sha-256;... URI.
+message ID {
+  // The parent ID, if any.
+  ID parent = 1;
+
+  // The GraphQL type of the value.
+  Type type = 2;
+
+  // GraphQL field name.
+  string field = 3;
+
+  // GraphQL field arguments, always in alphabetical order.
+  repeated Argument args = 4;
+
+  // If true, this Selector is not reproducible.
+  //
+  // TODO: do we need to refer to session/client IDs or anything here? Or is
+  // that all internal? Forcing function is whether this is used as an
+  // in-memory query cache key. But the query cache might be made per-session
+  // or even per-client instead anyway! What buys us the most?
+  bool tainted = 5;
+
+  // If true, this Selector may be omitted from the pipeline without changing
+  // the ultimate result.
+  //
+  // This is used to prevent meta-queries like 'pipeline' and 'withFocus' from
+  // busting cache keys when desired.
+  //
+  // It is worth noting that we don't store meta information at this level and
+  // continue to force metadata to be set via GraphQL queries. It makes IDs
+  // always easy to evaluate.
+  bool meta = 6;
+
+  // If the field returns a list, this is the index of the element to select.
+  // Note that this defaults to zero, as IDs always refer to
```

Comment:
I think the comment is cut off here maybe? Which left me waiting in anticipation because I was wondering about whether `Enumerable` values were indexed starting at `1` or `0` 😄

Reply:
lol, yes unfortunately they are indexed starting with `1`. :joy: I regret that decision, we should probably undo it.